src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -956,82 +1445,64 @@ ColocationIdForNewTable(Oid relationId, Var *distributionColumn,
 		 */
 		Assert(distributionMethod == DISTRIBUTE_BY_HASH);
 
-		Relation pgDistColocation = table_open(DistColocationRelationId(), ExclusiveLock);
-
 		Oid distributionColumnType = distributionColumn->vartype;
 		Oid distributionColumnCollation = get_typcollation(distributionColumnType);
-		bool createdColocationGroup = false;
 
-		if (pg_strncasecmp(colocateWithTableName, "default", NAMEDATALEN) == 0)
-		{
-			/* check for default colocation group */
-			colocationId = ColocationId(shardCount, ShardReplicationFactor,
-										distributionColumnType,
-										distributionColumnCollation);
+		/* get an advisory lock to serialize concurrent default group creations */
+		AcquireColocationDefaultLock();
 
-			/*
-			 * if the shardCount is strict then we check if the shard count
-			 * of the colocated table is actually shardCount
-			 */
-			if (shardCountIsStrict && colocationId != INVALID_COLOCATION_ID)
-			{
-				Oid colocatedTableId = ColocatedTableId(colocationId);
+		bool colocatedWithDefault = false;
+		colocationId = FindColocateWithColocationId(relationId,
+													replicationModel,
+													distributionColumnType,
+													distributionColumnCollation,
+													shardCount,
+													shardCountIsStrict,
+													colocateWithTableName,
+													&colocatedWithDefault);
```

Comment:
I don't think we really need this complex function to tell us whehter we're dealing with colocate_with := 'default'. We can just call IsColocateWithDefault directly.